src/charm.py

```diff
@@ -181,44 +173,47 @@ def _configure_amf(self, event: EventBase) -> None:  # noqa C901
                 return
         if not self._database_is_available():
             self.unit.status = WaitingStatus("Waiting for the amf database to be available")
-            event.defer()
             return
         if not self._get_database_info():
             self.unit.status = WaitingStatus("Waiting for AMF database info to be available")
-            event.defer()
             return
         if not self._nrf_requires.nrf_url:
             self.unit.status = WaitingStatus("Waiting for NRF data to be available")
-            event.defer()
             return
         if not self._amf_container.exists(path=CONFIG_DIR_PATH):
             self.unit.status = WaitingStatus("Waiting for storage to be attached")
-            event.defer()
             return
         if not _get_pod_ip():
             self.unit.status = WaitingStatus("Waiting for pod IP address to be available")
-            event.defer()
             return
-        if not self._certificate_is_stored():
+        if not self._private_key_is_stored():
+            self._generate_private_key()
+
+        if not self._csr_is_stored():
+            self._request_new_certificate()
             self.unit.status = WaitingStatus("Waiting for certificates to be stored")
-            event.defer()
             return
+
+        restart = False
+        csr = self._get_stored_csr()
+        for provider_certificate in self._certificates.get_assigned_certificates():
+            if provider_certificate.csr == csr:
+                existing_certificate = (
+                    self._get_stored_certificate() if self._certificate_is_stored() else ""
+                )
+                if not existing_certificate == provider_certificate.certificate:
+                    self._store_certificate(certificate=provider_certificate.certificate)
+                    restart = True
```

Comment:
this could be extracted in a method. the complexity of `_configure_amf` is already too high

Reply:
Done